Automattic/wp-calypso
# Editor Media Google Photos: date range filter

This is part of the wider migration from Picassa to Google Photos as per https://github.com/Automattic/wp-calypso/issues/29965.

## Description

Several issues have been raised about the difficulty of locating specific photos within external media, specifically Google Photos.

This PR introduces a mechanism to allow users to filter to show photos within a given date range. This will allow for users to easily jump back in time to locate a _specific_ folder and avoid the endless scrolling that is currently required to retrieve older photos.

## Important Notes

The `DateRange` component from this PR ~is being worked on separately within a new~ has been deprecated in favour of the core component created in this PR https://github.com/Automattic/wp-calypso/pull/29938.

Also see https://github.com/Automattic/wp-calypso/pull/30380

## Features
The date range filter will have the following features:

- [x] ability to select a date range (start - end) 
- [x] add a `dateFrom` and `dateTo` field to the API query to allow WPCom to make API  request for photos which fall within this range

## Test Plan

### Pre-testing prepation steps
You will need to be using the new Google Photos API integration to test this branch. In order to do this:

1. Visit the Sharing page and disconnect from the existing Google Photos (Picasa)
2. Apply the API patch in `D23306-code` 
3. Sandbox `public-api.wordpress.com`

Once done take the following steps:

1. Visit the Sharing page (you should _already_ be disconnected from Picassa Google Photos - if not go back and start again)
2. Connect to Google Photos - if you followed the steps above correctly you will see a oAuth dialogue which clearly says "Google Photos" and shows the logo. If you see anything else then you did it wrong! 
![screen shot 2019-01-25 at 16 58 07](https://user-images.githubusercontent.com/444434/51760530-834b3880-20c2-11e9-9ad7-60f984057ada.png)
3. Visit the Media modal and select Google Photos from the media library dropdown (top left of modal)
4. Verify that you can paginate through results ok.

You are now ready to start testing 😓...

### Testing

__Stop__: have you followed the preparation steps (above)? If not please do that _first_ then return here.

We are testing that you can filter your Google Photos library by a given date range using the UI provided. You will need to manually verify what are the appropriate date ranges of photos from you Google Library to test against. You will need:

* A known date range where you _do_ have photos
* A known date range where you do _not_ have photos (this can be very small, eg: 1-2 days)

Once you have this then you will need repeat the following steps anywhere where the Media Modal appears. So far we believe you will need test both of the following instances:

1. Post editor - go to any post and click "Featured Image" to open the Media Modal
2. Media Library - click on `Media` in the LHS menu in Calypso 

With the Modal open, verify the following scenarios (note you can edit and copy the following checkboxes when reporting on testing results in the comments):

- [ ] You can see photos with no dates selected
- [ ] You can filter a known date range and see the photos you expect to see (ie: the photos that fall within the selected dates)
- [ ] You can filter by a known date range where there are _not any photos_ and see the "No results" message displayed
- [ ] You can filter by a single date (ie: `from` only) and see dates from that date to `infinity`. If we have only 1 date then Google will assume the other end of the range is "infinity".
- [ ] With a range selected you can "clear" you date filter and go back to see all photos.

Repeat the above steps on:

- [ ] Firefox/Chrome (whichever is _not_ your usual browser)
- [ ] MS Edge
- [ ] Mobile device (making your brower window smaller _does not_ count, not does using Devtools device emulation mode!)
- [ ] Tablet device (making your brower window smaller _does not_ count, not does using Devtools device emulation mode!)

### Other Items to Test

* Resolution of eslint warnings (see https://github.com/Automattic/wp-calypso/pull/25762/commits/b081382d34a4d6cba37923c2b360bfa96c92d268)
* Fix of string-based `ref` usage in DatePicker https://github.com/Automattic/wp-calypso/pull/25762/commits/41ae8629b83fbacc81be6e74aaf9313c8f92e92b

## Todo
- [ ] Remove usages such as `transparentize( $gray-lighten-20, 0.5 )`.
- [x] Rebase and ensure all styles are migrated to Webpack ES Imports ( see `p4TIVU-98Q-p2`) 
- [x] Display appropriate UI when no photos are found which match a Date filter. Current says `You don't have any photos in your Google library` which isn't true.
- [x] Determine most appropriate format to pass Dates (`Date`, Moment or ISO string) to `onQueryFilterChange` handler
- [x] Guard against only start or only end date scenarios. Should cope gracefully and fall back to "infinity". 
- [x] Remove legacy DateRange styles from Media Library styles.scss
- [x] Remove legacy MediaDateRange component
- [x] Add ability to full clear the date filters and go back to "All dates" - needs guard against `InvalidDate-InvalidDate` scenario (being worked on in https://github.com/Automattic/wp-calypso/pull/30380)
- [x] review [merge checklist docs](https://github.com/Automattic/wp-calypso/blob/master/docs/merge-checklist.md)
- [x] UX on smaller viewport sizes
- [x] ~refactor `MediaDateRange` snapshot test to use more explicit assertions about shape~
- [x] ~pull in existing `master` of `package.json` and `npm-shrinkwrap` to resolve conflict~
- [x] ~add input fields within popover to afford manual entry for those who cannot (or prefer not to) utilise the datepicker interface~


## Checklist:
